### PR TITLE
Add MultipartFileArgumentResolver for HTTP interface client

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**
- [x] add tests
- [x] add more javadocs

### CODE REVIEW

**Comment by OlgaMaciaszek** [comment]
Fixes gh-30342.

File: spring-web/src/main/java/org/springframework/web/service/invoker/MultipartFileArgumentResolver.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	private static final String MULTIPART_FILE_LABEL = "multipart file";
+
+	@Override
+	protected NamedValueInfo createNamedValueInfo(MethodParameter parameter) {
+		if (!parameter.nestedIfOptional().getNestedParameterType().equals(MultipartFile.class)) {
+			return null;
+		}
+		String parameterName = parameter.getParameterName();
+		return new NamedValueInfo(parameterName != null ? parameterName : "", true,
+				null, MULTIPART_FILE_LABEL, true);
```
Comment (rstoyanchev) [inline spring-web/src/main/java/org/springframework/web/service/invoker/MultipartFileArgumentResolver.java:48]: You can pass the name as `""`. It's checked later on in the base class and updated.

File: spring-web/src/main/java/org/springframework/web/service/invoker/MultipartFileArgumentResolver.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		requestValues.addRequestPart(name, toHttpEntity(name, file));
+	}
+
+	private HttpEntity<Resource> toHttpEntity(String name, MultipartFile file) {
+		HttpHeaders headers = new HttpHeaders();
+		if (file.getOriginalFilename() != null) {
+			headers.setContentDispositionFormData(name, file.getOriginalFilename());
+		}
+		if (file.getContentType() != null) {
+			headers.setContentType(MediaType.parseMediaType(file.getContentType()));
```
Comment (rstoyanchev) [inline spring-web/src/main/java/org/springframework/web/service/invoker/MultipartFileArgumentResolver.java:70]: We can avoid parsing here, and use `headers.addHeader(HttpHeaders.CONTENT_TYPE, file.getContentType())`.

File: spring-web/src/main/java/org/springframework/web/service/invoker/MultipartFileArgumentResolver.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+	}
+
+	@Override
+	protected void addRequestValue(String name, Object value, MethodParameter parameter,
+			HttpRequestValues.Builder requestValues) {
+		Assert.isInstanceOf(MultipartFile.class, value,
+				"The value has to be of type 'MultipartFile'");
+		Assert.isInstanceOf(MultipartFile.class, value,
+				"The value has to be of type 'MultipartFile'");
```
Comment (rstoyanchev) [inline spring-web/src/main/java/org/springframework/web/service/invoker/MultipartFileArgumentResolver.java:58]: `Assert.state` would be more appropriate here, `IllegalStateException` vs `IllegalArgumentException`.

**Review by rstoyanchev — CHANGES_REQUESTED** [review]
Overall looks good. Some minor comments below.

**Review by rstoyanchev — APPROVED** [review]
Thanks for the updates, and for adding an integration test.

**Pull request closed without merging**